src/DynamoCoreWpf/Views/Core/NoteView.xaml.cs

```diff
@@ -152,5 +168,268 @@ private void PrepareZIndex()
+        private static int TAB_SPACING_SIZE = 4;
+
+        /// <summary>
+        /// Amount of spaces inserted after bullet
+        /// </summary>
+        private static int SPACING_AFTER_BULLET = 2;
+
+        /// <summary>
+        /// Handles text when user types Key.OemMinus (DASH)
+        /// It converts the DASH into its appropriate BULLET Character
+        /// by counting the amount of TAB spaces before the bullet.  
+        /// Conversion only happens when dash is at the beginning of a line.
+        /// </summary>
+        /// <param name="text"> text before typing DASH </param>
+        /// <param name="caretIndex"> caret index where user typed DASH </param>
+        /// <returns></returns>
+        private string BulletDashHandler( string text, int caretIndex)
+        {
+            // Get the line where DASH was typed 
+            var lineNumber = GetLineNumberAtCaretsIndex(text, caretIndex);
+            var line = GetLineTextAtCaretsIndex(text, caretIndex);
+            var caretAtLine = GetCaretIndexRelativeToLine(text, caretIndex);
+
+            // If there is text before caret
+            // or another bullet in line
+            // dont convert it to bullet
+            bool lineContainsBullet = BULLETS_CHARS.Where(b => line.Contains(b)).Any();
+            var textBeforeCaret = line.Substring(0, caretAtLine);
+            if (!StringIsEmptyOrTab(textBeforeCaret)&& !lineContainsBullet)
+            {
+                line = line.Insert(caretAtLine, "-");
+                return ReplaceLineOfText(text, lineNumber, line);
+            }
+
+            var tabsBeforeCaret = textBeforeCaret.Count(t => t == '\t');
+            var bulletIndex = tabsBeforeCaret % BULLETS_CHARS.Length;
+            var bulletCharAndRightSpacing = BULLETS_CHARS[bulletIndex] + new String(' ', SPACING_AFTER_BULLET);
+            line = line.Insert(caretAtLine, bulletCharAndRightSpacing);
+
+            return ReplaceLineOfText(text, lineNumber, line);
+        }
+
+        /// <summary>
+        /// Handles text when user types Key.Tab (TAB)
+        /// If TAB is pressed before a bullet or right after the bullet
+        /// it also updates the bullet to its appropriate bullet character
+        /// </summary>
+        /// <param name="text">  text before typing TAB </param>
+        /// <param name="caretIndex"> caret index where user typed TAB </param>
+        /// <returns></returns>
+        private string BulletTabHandler(string text, int caretIndex)
+        {
+            // Get the line where TAB was typed 
+            var lineNumber = GetLineNumberAtCaretsIndex(text, caretIndex);
+            var line =  GetLineTextAtCaretsIndex(text,caretIndex);
+            var caretAtLine = GetCaretIndexRelativeToLine(text, caretIndex);
+
+            // If TAB was pressed just after a bullet insert it before the bullet
+            // so that you can easily indent
+            var distanceFromBulletToCaret = SPACING_AFTER_BULLET + 1;
```

Comment:
Do we need to define this `1` as const somehow?

Reply:
@QilongTang the `SPACING_AFTER_BULLET` is the constant. I think `distanceFromBulletToCaret` doesn't need to be another constant as its always a variable dependant on `SPACING_AFTER_BULLET + 1`, 
It just adds 1 as the caret index is at the end of the character being typed.

Example:
   *_ _ -> two spaces
   *_ _ | -> caretIndex is two spaces + 1 to bullet index.